### PR TITLE
Opt into has many inverse

### PR DESCRIPTION
**Pull request by gmcgibbon**
### Summary

Followup on https://github.com/rails/rails/pull/34533, https://github.com/rails/rails/pull/37413

Make `has_many` inversing support available through an opt-in config variable. This behaviour is likely to break existing applications, but it is correct behaviour.

### CODE REVIEW

File: activerecord/lib/active_record/railtie.rb [hunk 1]
```diff
@@ -67,6 +68,12 @@ class Railtie < Rails::Railtie # :nodoc:
       require "active_record/base"
     end
 
+    initializer "active_record.has_many_inversing" do |app|
+      ActiveSupport.on_load(:active_record) do
+        self.has_many_inversing = app.config.active_record.has_many_inversing
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/railtie.rb]: You don't need this intializer. The setting config initializer will take care of it.

Comment (gmcgibbon) [reply activerecord/lib/active_record/railtie.rb]: Fixed. CI should also pass now

**Comment by kamipo** [comment]
We'd gradually improved automatic inverse_of detection without opt-in configuration (e.g. #28808, #28830).
Is that change (#34533) broken any app in Shopify?

**Comment by gmcgibbon** [comment]
@kamipo Yes, it did introduce breakages. I think it was related to two different cases:

1. Code that doesn't expect the inverse to be set on `has_many` relations (I think this counts as a behaviour change in Active Record, so I'm choosing to make it opt-in)
2. Associations that use `before_add` or `after_add` (fixed in #37413)

See https://github.com/rails/rails/pull/34533#issuecomment-539225561

**Comment by eileencodes** [comment]
@gmcgibbon and @kamipo is this ready to be merged? We're hitting this issue at GitHub too 😄

**Comment by gmcgibbon** [comment]
@eileencodes Yep, it should be ready to merge. I'm just waiting for reviews.

**Review by rafaelfranca — APPROVED** [review]

**Pull request merged**